### ISSUE TITLE
IDNA checks in BASIC00

### ISSUE THREAD

**Issue by mattias-p** [dotse/zonemaster#614]
* A message MUST be emitted for IDNA-invalid A-labels.
* A message MUST be emitted for IDNA-forbidden code points encoded in A-labels.

**Comment by pawal** [comment]
DNS is essentially 8-bit clean, and does not internally do any conversion of IDNA anywhere. It's pure application logic. Any application that uses Zonemaster can be free to do this kind of tests, but the Engine should stay away from it. I have very strong opinions about Zonemaster being a DNS delegation test tool, and not a tool for checking things that is outside of the DNS protocol.

IDNA is an abbreviation of "Internationalized Domain Names for Applications".

If you insist, please make these tests optional.

**Comment by mattias-p** [comment]
Yeah, I doubt anyone thinks these messages should default to the ERROR severity level. NOTICE seems appropriate.

**Comment by matsduf** [comment]
This issue specifies that *messages* must be emitted, not change of requirements. The message is one thing, the severity level is another. The latter is set in the profile/policy. Messages on errors in A-labels can be set to INFO or even DEBUG by the application using Zonemaster.

We have started a work to make the messages more integrated in the test cases. See dotse/zonemaster#609.

Basic00 already states:

> In order to begin testing the domain name from the input must be a valid
> domain name. The domain name must either be a 
> 
> 1. a valid IDN name (Internationalized Domain Name) ([RFC 5890](https://tools.ietf.org/html/rfc5890#page-13), 
> section 2.3.2.3), or
> 2. a valid ASCII domain.
> 
> The ASCII domain name is valid if follows the rules defined in section 2.1 
> of [RFC 1123](https://tools.ietf.org/html/rfc1123#section-2.1), i.e. only 
> consists of the ASCII characters "a-z", "A-Z", "0-9", "." and "-" with the 
> extension of the "_" character, standardized for SRV records 
> ([RFC 2782](https://tools.ietf.org/html/rfc2782)). The "." character is 
> the delimiter between labels, and a label must not start or end with a 
> "-" character. There are also lenght limitation of the domain names and 
> its labels (specified below). 
>

**Comment by pawal** [comment]
But then BASIC00 is perhaps the wrong place for it. All names used by the domain name should be tested, such as name servers. The purpose of the BASIC tests is to see if it is possible to perform the tests.

I can quote the test plan for the BASIC tests:

> These are tests of a domain's most basic functionality. If these fail, it will be hard or impossible to perform any other tests at all. The test code should be constructed so that these tests are always run, even if a subset of tests is asked for that would not normally include them.

**Comment by matsduf** [comment]
This issue only asks for explicit messages for tests that are already part of Basic00. Do you have a suggestion for where IDNA tests of the domain name should be instead?

Please note that, as we see it, the test cases cover all parts of Zonemaster, not only Zonemaster-Engine.

Since Zonemaster is configurable, it is possible to put your own severity on the messages emitted.

**Comment by pawal** [comment]
SYNTAX is the obvious place to put these.

**Comment by matsduf** [comment]
Good suggestion. I can see that we already have a test case there covering what is also in Basic00. In Zonemaster-Backend we want to perform some tests early to stop some wrong input data from even intiating a test (early error), and then Basic00 has been a natural candidate, but there is nothing that prevents Syntax01 to be run early together with another SyntaxZZ test case for IDNA names.

**Comment by pawal** [comment]
There is no problem letting the Engine test illegal IDNA names, they are legal DNS names. If a UI wants to stop the Engine from testing illegal IDNA names, that logic belongs to the application.

**Comment by matsduf** [comment]
Different applications can have different severity levels on the messages. Zonemaster-Engine is not run on its own, it is always part of an application. The specification is for Zonemaster as a whole, not just for Zonemaster-Engine.

**Comment by pawal** [comment]
If you don't want to run a test based on the domain name, that logic fits better in the application. There are many good libraries that checks the validity of IDNA. Let me iterate this, IDNA is not DNS.

**Comment by matsduf** [comment]
PR #942 proposes an update of Basic00, but is does not follow this issue. If a names comes in as an A-label, it is probably no reason to stop it, but a test in Syntax could verify then name as this issue requires.
